btree/btree.c

```diff
@@ -1708,13 +1725,42 @@ static int64_t node_get(struct node_op *op, struct td *tree,
 	nt = &fixed_format;
 	op->no_node = nt->nt_opaque_get(addr);
 	if (op->no_node != NULL &&
-	    op->no_node->n_addr.as_core == addr->as_core)
-		op->no_node->n_ref++;
-	else {
+	    op->no_node->n_addr.as_core == addr->as_core) {
+		if (op->no_node->n_ref == 0) {
+			/**
+			 * The node descriptor is in LRU list. Remove from lru
+			 * list and add to trees active list
+			 */
+			node_refcnt_update(op->no_node, true);
+
+			m0_rwlock_write_lock(&lru_lock);
+			M0_ASSERT_EX(ndlist_tlist_contains(&btree_lru_nds,
+							   op->no_node));
+			ndlist_tlist_del(op->no_node);
+			m0_rwlock_write_unlock(&lru_lock);
+
+			m0_rwlock_write_lock(&tree->t_lock);
+			ndlist_tlist_add(&tree->t_active_nds, op->no_node);
+			m0_rwlock_write_unlock(&tree->t_lock);
+			/**
+			 * Update nd::n_tree  to point to tree descriptor as we
+			 * as we had set it to NULL in node_put(). For more
+			 * details Refer comment in node_put().
+			 */
+			op->no_node->n_tree = tree;
+		} else
+			node_refcnt_update(op->no_node, true);
+	} else {
 		/**
 		 * If node descriptor is not present allocate a new one
-		 * and assign to node. */
+		 * and assign to node.
+		 */
 		node = m0_alloc(sizeof *node);
+		/**
```

Comment:
What if 2 threads reach this same line at the same time on different CPUs !!!
Currently we do not have any protection from allocating 2 node descriptors for the same node !!!

Reply:
Added nd::n_lock to protect node descriptor field update.

Reply:
Even this does not help because both the threads would still allocate nodes because they have no way to know that other thread has already allocated. Need to think about this corner case. This is the time where Nikita mentioned about a global lock which can be used here. If you do not want to introduce a new lock you can maybe use the lru lock for now.

Reply:
Oh ok. Yes, will add it under lru lock for now and put a comment to have a separate global lock once the hash functionality is implemented.